### PR TITLE
MQTT-TLS documentation

### PR DESCRIPTION
**Pull request by alexandre-vasc**
This is a partial documentation about MQTT with TLS.
Sorry, I haven´t had the time to finish it.

This PR is connected to dojot/dojot#166

### CODE REVIEW

**Comment by giovannicuriel** [comment]
Can I finish this? I'll assign to myself.

File: source/security/mqtt-tls.rst [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+For a device to connect using TLS with Mosquitto, it must posses:
+  - A key pair (.key file)
+  - A certificate signed by a Certificate Authority (CA) trusted by Mosquitto (.crt file)
+  - The certificate of this CA (.crt file)
+  - An entry on Mosquitto Access Control List (ACL), allowing the device to publish on a specific topic.
+  - (optional and not so necessary) A Certificate Revocation List (CRL)
+
+
+When a device is created, Device-manager will (automatic) configure the following components:
```
Comment (mmagr) [inline source/security/mqtt-tls.rst]: manager will (~automatic~automatically) configure the

File: source/security/mqtt-tls.rst [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  - IoTAgent -> will registry the new device on its internal cache.
```
Comment (mmagr) [inline source/security/mqtt-tls.rst]: IoTAgent -> will ~registry~ **register** the new

File: source/security/mqtt-tls.rst [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  - MQTT-Manager -> Will create an entry on the ACL. Allowing the device to publish on a specific topic.
```
Comment (mmagr) [inline source/security/mqtt-tls.rst]: Perhaps:

MQTT-Manager -> Will create an entry on the ACL, allowing the device to publish on a specific topic.

File: source/security/mqtt-tls.rst [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  - EJBCA -> Will create an end entity so a certificate can be created on the future.
+
+By default, dojot use clear MQTT. To activate TLS, some changes must be done on docker-compose.
```
Comment (mmagr) [inline source/security/mqtt-tls.rst]: By default, dojot ~use~ **uses** clear MQTT

File: source/security/mqtt-tls.rst [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+On docker-compose.yml
+  - Change the image of the service 'mqtt' from 'ansi/mosquitto' to 'dojot/mqtt-manager'
+  - Change the avalible door for 'mqtt' service from '1883:1883' to '8883:8883'
```
Comment (mmagr) [inline source/security/mqtt-tls.rst]: Change the ~avalible door~ **public port**  for 'mqtt' service from '1883:1883' to '8883:8883'

File: source/security/mqtt-tls.rst [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+   - set the flag 'secure' to true
+
+Components
+----------
+
+EJBCA-REST
+***************
+
+EJBCA (https://www.ejbca.org) is a complete Private Key Infrastructure (PKI) capable to manage  CAs, cryptography keys and certificates.
+EJBCA provides a SOAP, web and a command line interface. EJBCA-REST is an wrapper on top of EJBCA that provides modern interfaces, like REST and Kafka.
```
Comment (mmagr) [inline source/security/mqtt-tls.rst]: perhaps?

EJBCA provides SOAP, web and command line interfaces. EJBCA-REST is a wrapper on top of EJBCA that complements those, allowing the CA to be configured using REST. When used within dojot, it also listens to kafka events, allowing its automatic configuration.

File: source/security/mqtt-tls.rst [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+What is a certificate?
+======================
+
+
+A certificate contains the public key for a entity (a user, device, website), together with a bunch of information about this entity, about the CA signing the certificate, the allowed certificate usage and a checksum.
+When a entity want a certificate signed, the entity should create a CSR file and sent it to the desired CA.
```
Comment (mmagr) [inline source/security/mqtt-tls.rst]: When ~a~ **an** entity ~want~ **wants** a certificate signed

File: source/security/mqtt-tls.rst [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+The CSR file is an 'intention of certification'. The fie contains the information required from the entity and some information about the certificate use, hostnames and IPs where the certificate will reside, alternative names for the entity and etc.
+EJBCA can decide, using its configured policies, what information to keep, to discard and to overwrite of the received CSR. EJBCA can refuse to sign a CSR not judged safe enough be it policies.
```
Comment (mmagr) [inline source/security/mqtt-tls.rst]: refuse to sign a CSR **judged not safe enough for its policies**.

File: source/security/mqtt-tls.rst [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+These configurable policies are called 'Certificate Profiles'. One Certificate profile named CFREE, specialized for MQTT TLS,  is provided out of the box.
+
+In short, CFREE have the following configurations (and many more):
+  - Cryptography keys must have between 2048 and 8192 bits.
+  - Certificate expires is 730 days.
```
Comment (mmagr) [inline source/security/mqtt-tls.rst]: **Certificates expire in 730 days.**

File: source/security/mqtt-tls.rst [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  - Key usage is marked as not critical (for now)
+  - The hash algorithm is SHA256. The sign algorithm is RSA.
+
+
+I have a CSR. How can I ask EJBCA to sign it for me?
+====================================================
+
+Calm down! EJBCA will not allow strangers to ask for certification. You need to authenticate yourself.
+EJBCA use a username+password authentication system. The term 'end entity' will be used to refer to EJBCA users to follow the terms on EJBCA documentation and to avoid ambiguities between EJBCA users and dojot users.
+An administrator should create the end entity. A just created end entity have the state 'New' an can generate a certificate. After sign a certificate for an entity, the end entity pass to the state 'Generated' and will no longer accept this username and password.
```
Comment (mmagr) [inline source/security/mqtt-tls.rst]: A just created end entity ~have~ **has** the state 'New'

File: source/security/mqtt-tls.rst (hunk 1 above)
Comment (giovannicuriel) [reply source/security/mqtt-tls.rst]: Ok

File: source/security/mqtt-tls.rst (hunk 2 above)
Comment (giovannicuriel) [reply source/security/mqtt-tls.rst]: Ok

File: source/security/mqtt-tls.rst (hunk 3 above)
Comment (giovannicuriel) [reply source/security/mqtt-tls.rst]: Ok

File: source/security/mqtt-tls.rst (hunk 5 above)
Comment (giovannicuriel) [reply source/security/mqtt-tls.rst]: Ok

File: source/security/mqtt-tls.rst (hunk 6 above)
Comment (giovannicuriel) [reply source/security/mqtt-tls.rst]: Ok

File: source/security/mqtt-tls.rst (hunk 8 above)
Comment (giovannicuriel) [reply source/security/mqtt-tls.rst]: Ok

File: source/security/mqtt-tls.rst (hunk 10 above)
Comment (giovannicuriel) [reply source/security/mqtt-tls.rst]: Ok

**Comment by giovannicuriel** [comment]
This PR is connected to dojot/dojot#166

File: source/security/mqtt-tls.rst [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+- SSL alert number 45: this error indicates that a certificate expired. Keep in
+  mind that CRLs also expire.
+- SSL alert number 48: received a valid certificate chain or partial chain, 
+  but the certificate was not accepted because the CA certificate could not be 
+  located or could not be matched with a known, trusted CA. This message is 
+  always fatal.
+
+- Alert unknown CA: check whether sent CA certificate is correct. If it is a
+  sub-CA, check if all of its certificate chain was sent. This error also occurs
+  if the CA certificate data (specially common name attribute) is the same as
```
Comment (mmagr) [inline source/security/mqtt-tls.rst]: is this right? shouldn't it read:

> if the CA certificate data (specially common name attribute) is **not** the same as those from client certificate

Comment (alexandre-vasc) [reply source/security/mqtt-tls.rst]: you are right @mmagr

**Pull request merged**